### ISSUE TITLE
Introduce new `suspend` and rename

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#891]
Currently, `ZIO.suspend` assumes the effect is total, but that's not necessarily the case.

We should have:

 - `ZIO.effectSuspendWith(p: Platform => TaskR[...]): TaskR[...]`
 - `ZIO.effectSuspendTotalWith(p: Platform => ZIO[...]): ZIO[...]`
 - `ZIO.effectSuspend(taskr: => TaskR[...]): TaskR[...]`
 - `ZIO.effectSuspendTotal(zio: => ZIO[...]): ZIO[...]`

This matches the convention for `effect` / `effectTotal` and will be a good default choice (I saw code in the wild using `suspend` and incorrectly assuming partial effects).

This will require 1 new operation added (`SuspendTotal`), co-opting the existing `Suspend` into the partial suspend, and changing type signatures as appropriate.

In addition, the operations should be available in `IO`, `Task`, `TaskR`, and the pure ones (`Total`) can be available in `UIO`.

**Comment by kamilkloch** [comment]
I would like to pick up this issue.

**Comment by kamilkloch** [comment]
> We should have:
> 
> * `ZIO.suspendWith(p: Platform => TaskR[...]): TaskR[...]`
> * `ZIO.suspendTotalWith(p: Platform => ZIO[...]): ZIO[...]`
> * `ZIO.suspend(taskr: => TaskR[...]): TaskR[...]`
> * `ZIO.suspendTotal(zio: => ZIO[...]): ZIO[...]`

@jdegoes, could you perhaps help to clarify? Following the `effect` / `effectTotal` convention I came up with the following signatures...
```scala
suspendWith[R >: LowerR, E <: UpperE, A](p: Platform => ZIO[R, E, A]): ZIO[R, E, A] 
suspendTotalWith[A](p: Platform => UIO[A]): UIO[A]
suspend[R >: LowerR, E <: UpperE, A](zio: => ZIO[R, E, A]): ZIO[R, E, A]
suspendTotal[A](uio: => UIO[A]): UIO[A]
``` 
Thanks!

**Comment by jdegoes** [comment]
The correct signatures will be as follows:

```scala
suspendWith[A](p: Platform => Task[A]): Task[A] 
suspend[A](task: => Task[A]): Task[A] = suspendWith(_ => task)

suspendTotalWith[A](p: Platform => UIO[A]): UIO[A]
suspendTotal[A](uio: => UIO[A]): UIO[A] = suspendTotalWith(_ => uio)
```

The `suspendTotal[With]` constructors (the functions passed to these functions) are _not_ allowed to throw exceptions; while the `suspend[With]` constructors (the functions passed to these functions) _are_ allowed to throw `Throwable`, and such `Throwable` will be reflected in the return type.

You can see the similar distinction between `effect` and `effectTotal` today.

Does this make sense? 😄

**Comment by kamilkloch** [comment]
It does, partially :) Agreed on `suspendTotal ` and `suspendTotalWith`.
I understand why `effect` produces `Task[A]`(no environment available), but as for `suspend` wouldn't a more general signature look like `suspend[R >: LowerR, A](taskr: => TaskR[R, A]): TaskR[R, A]`?

**Comment by jdegoes** [comment]
You are right! Also we could do ZIO[R, Nothing, A] for the other one (total variants).

**Comment by kamilkloch** [comment]
OK, thanks for a quick clarification :)

**Comment by kamilkloch** [comment]
`ZIO.suspend` is currently used in places with a generic error type `E`. Example: `FunctionIO.scala`:215-223:
```scala
  private[zio] final class Impure[E, A, B](val apply0: A => B) extends FunctionIO[E, A, B] {
    val run: A => IO[E, B] = a =>
      IO.suspend {
        try IO.succeed[B](apply0(a))
        catch {
          case e: FunctionIOError[_] => IO.fail[E](e.unsafeCoerce[E])
        }
      }
  }
```
The proposed new (restictive) signature of `ZIO.suspend`
```scala
suspend[R, A](task: => TaskR[R, A]): TaskR[R, A] 
```
does not cut it. 

Conceptually, keeping the current implementation
```scala
suspend[R, E, A](zio: => ZIO[R, E, A]): ZIO[R, E, A]
```
makes more sense to me.

**Comment by jdegoes** [comment]
For consistency, we need to prefix these methods by `effect`, e.g. `effectSuspend`.

This means deprecating existing `suspend`.

**Comment by kamilkloch** [comment]
I am not sure if #926 is right (with regard to the error type). Current
```scala
effectSuspend(zio: => ZIO[R, E, A]): ZIO[R, E, A]
```
is as generic as possible, but `effectSuspend`handles `throwables`, not generic `E`s. 

On the other hand, 
```scala
effectSuspend[R, A](task: => TaskR[R, A]): TaskR[R, A]
```
does not work for current `FunctionIO.Impure` implemenation (generic `E`):
```scala
private[zio] final class Impure[E, A, B](val apply0: A => B) extends FunctionIO[E, A, B] {
    val run: A => IO[E, B] = a =>
      IO.suspend {
        try IO.succeed[B](apply0(a))
        catch {
          case e: FunctionIOError[_] => IO.fail[E](e.unsafeCoerce[E])
        }
      }
  }
```

**Comment by ghostdogpr** [comment]
Closed by #926
